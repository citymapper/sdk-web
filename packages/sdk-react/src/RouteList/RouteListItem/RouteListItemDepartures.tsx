import type {
  Departure,
  DepartureTransitLive,
  Stop,
} from '@citymapper/api/@types'
import React from 'react'
import styled from '../../utils/cmStyled'
import DepartureFrequency from './Departures/DepartureFrequency'
import DepartureLive from './Departures/DepartureLive'
import DeparturesScheduled from './Departures/DeparturesScheduled'

export interface RouteListItemDeparturesProps {
  departures: Departure[]
  stop: Stop
}

const Wrapper = styled('div', { className: 'cm-route-departures' })`
  font-size: 0.8125em;
  flex: 1;
  margin-top: ${({ theme }) => theme.spacing(1.75)};
  &.cm-route-departures {
    ${({ theme }) => theme.CmDepartures?.root}
  }
`

const RouteListItemDepartures: React.FC<RouteListItemDeparturesProps> = ({
  departures,
  stop,
}) => {
  if (!departures || departures.length === 0) return null

  switch (departures[0].type) {
    case 'frequency': {
      return (
        <Wrapper data-departure-type="frequency">
          <DepartureFrequency
            frequencyRange={departures[0].frequency_seconds_range}
            originStation={stop.name}
          />
        </Wrapper>
      )
    }
    case 'scheduled': {
      if (!departures[0].scheduled_time) return null
      return (
        <Wrapper data-departure-type="scheduled">
          <DeparturesScheduled
            time={departures[0].scheduled_time}
            status={departures[0].time_status}
            headSign={departures[0].headsign}
            originStation={stop.name}
          />
        </Wrapper>
      )
    }
    case 'live': {
      const liveDepartures = departures.filter(
        (d) => d.type === 'live'
      ) as DepartureTransitLive[]
      return (
        <Wrapper data-departure-type="live" data-is-live="true">
          <DepartureLive
            departures={liveDepartures}
            originStation={stop.name}
          />
        </Wrapper>
      )
    }
    default:
      return null
  }
}

export default RouteListItemDepartures
