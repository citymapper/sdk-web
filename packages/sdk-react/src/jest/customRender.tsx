import { render, RenderOptions } from '@testing-library/react'
import React, { JSXElementConstructor, ReactElement } from 'react'
import CMProvider from '../CMProvider'

const TestWrapper: React.FC<{ children: React.ReactElement }> = ({
  children,
}) => (
  <CMProvider
    config={{
      apiKey: '',
      citymapperFont: false,
    }}
  >
    {children}
  </CMProvider>
)

const customRender = (
  ui: ReactElement<any, string | JSXElementConstructor<any>>,
  options?: RenderOptions<
    typeof import('@testing-library/dom/types/queries'),
    HTMLElement,
    HTMLElement
  >
) => render(ui, { wrapper: TestWrapper, ...options })

export default customRender
