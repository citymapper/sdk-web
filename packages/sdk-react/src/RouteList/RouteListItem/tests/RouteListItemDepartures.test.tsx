import { DateTime } from 'luxon'
import React from 'react'
import customRender from '../../../jest/customRender'
import RouteListItemDepartures from '../RouteListItemDepartures'

describe('RouteListItemDepartures', () => {
  it('frequency: single departure is displayed correctly', () => {
    const { container } = customRender(
      <RouteListItemDepartures
        stop={{
          name: 'Stratford',
          coordinates: { lat: 0, lon: 0 },
          id: 'Stratford',
          indicator_text: '',
        }}
        departures={[
          {
            type: 'frequency',
            service_id: 'LondonBus89',
            frequency_seconds_range: [720, 720],
            time_status: 'on_time',
          },
        ]}
      />
    )

    expect(container).toMatchSnapshot()
  })

  it('frequency: departure range is displayed correctly', () => {
    const { container } = customRender(
      <RouteListItemDepartures
        stop={{
          name: 'Stratford',
          coordinates: { lat: 0, lon: 0 },
          id: 'Stratford',
          indicator_text: '',
        }}
        departures={[
          {
            type: 'frequency',
            service_id: 'LondonBus89',
            frequency_seconds_range: [720, 800],
            time_status: 'on_time',
          },
        ]}
      />
    )

    expect(container).toMatchSnapshot()
  })

  it('scheduled: single departure is displayed correctly', () => {
    const { container } = customRender(
      <RouteListItemDepartures
        stop={{
          name: 'Stratford',
          coordinates: { lat: 0, lon: 0 },
          id: 'Stratford',
          indicator_text: '',
        }}
        departures={[
          {
            type: 'scheduled',
            service_id: 'NationalRailSE',
            headsign: 'Dartford',
            scheduled_time: '2022-06-30T14:00:00+01:00',
            time_status: 'on_time',
          },
        ]}
      />
    )

    expect(container).toMatchSnapshot()
  })

  it('live: single departure display correctly', () => {
    const { container } = customRender(
      <RouteListItemDepartures
        stop={{
          name: 'Stratford',
          coordinates: { lat: 0, lon: 0 },
          id: 'Stratford',
          indicator_text: '',
        }}
        departures={[
          {
            type: 'live',
            service_id: 'Central',
            headsign: 'Ealing Broadway',
            live_time: DateTime.now().plus({ minute: 4 }).toISO(),
            scheduled_time: DateTime.now().plus({ minute: 4 }).toISO(),
          },
        ]}
      />
    )

    expect(container).toMatchSnapshot()
  })

  it('live: multiple departures display correctly', () => {
    const { container } = customRender(
      <RouteListItemDepartures
        stop={{
          name: 'Stratford',
          coordinates: { lat: 0, lon: 0 },
          id: 'Stratford',
          indicator_text: '',
        }}
        departures={[
          {
            type: 'live',
            service_id: 'Central',
            headsign: 'Ealing Broadway',
            live_time: DateTime.now().plus({ minute: 4 }).toISO(),
            scheduled_time: DateTime.now().plus({ minute: 4 }).toISO(),
          },
          {
            type: 'live',
            service_id: 'Central',
            headsign: 'Ealing Broadway',
            live_time: DateTime.now().plus({ minute: 9 }).toISO(),
            scheduled_time: DateTime.now().plus({ minute: 9 }).toISO(),
          },
          {
            type: 'live',
            service_id: 'Central',
            headsign: 'Ealing Broadway',
            live_time: DateTime.now().plus({ minute: 15 }).toISO(),
            scheduled_time: DateTime.now().plus({ minute: 15 }).toISO(),
          },
        ]}
      />
    )

    expect(container).toMatchSnapshot()
  })
})
