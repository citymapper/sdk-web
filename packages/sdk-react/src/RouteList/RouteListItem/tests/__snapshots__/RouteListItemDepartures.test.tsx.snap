// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteListItemDepartures frequency: departure range is displayed correctly 1`] = `
.emotion-0 {
  font-size: 0.8125em;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-top: 14px;
}

.emotion-3 {
  color: #727780;
}

<div>
  <div
    class="cm-route-departures emotion-0"
    data-departure-type="frequency"
  >
    <div
      class="cm-departure-frequency-root emotion-1"
    >
      <span
        class="cm-departure-frequency-departures emotion-2"
      >
        Every 
        12-14
         min
      </span>
       
      <span
        class="cm-departure-frequency-origin emotion-3"
      >
        from 
        Stratford
      </span>
    </div>
  </div>
</div>
`;

exports[`RouteListItemDepartures frequency: single departure is displayed correctly 1`] = `
.emotion-0 {
  font-size: 0.8125em;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-top: 14px;
}

.emotion-3 {
  color: #727780;
}

<div>
  <div
    class="cm-route-departures emotion-0"
    data-departure-type="frequency"
  >
    <div
      class="cm-departure-frequency-root emotion-1"
    >
      <span
        class="cm-departure-frequency-departures emotion-2"
      >
        Every 
        12
         min
      </span>
       
      <span
        class="cm-departure-frequency-origin emotion-3"
      >
        from 
        Stratford
      </span>
    </div>
  </div>
</div>
`;

exports[`RouteListItemDepartures live: multiple departures display correctly 1`] = `
.emotion-0 {
  font-size: 0.8125em;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-top: 14px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: flex-start;
  align-self: flex-start;
  margin-right: 4px;
}

.emotion-2 svg {
  width: 8px;
  height: 8px;
}

.emotion-3 {
  width: 24px;
  height: 24px;
}

.emotion-5 {
  color: #727780;
}

<div>
  <div
    class="cm-route-departures emotion-0"
    data-departure-type="live"
    data-is-live="true"
  >
    <div
      class="cm-departure-live-root emotion-1"
    >
      <div
        class="cm-departure-live-blip emotion-2"
      >
        <svg
          class="emotion-3"
          fill="none"
          viewBox="0 0 8 8"
        >
          <path
            clip-rule="evenodd"
            d="M3.46447 3.46447C4.40215 2.52678 5.67392 2 7 2C7.55228 2 8 1.55228 8 1C8 0.447715 7.55228 0 7 0C5.14348 0 3.36301 0.737498 2.05025 2.05025C0.737498 3.36301 0 5.14348 0 7C0 7.55228 0.447715 8 1 8C1.55228 8 2 7.55228 2 7C2 5.67392 2.52678 4.40215 3.46447 3.46447ZM5.93934 5.93934C6.22064 5.65804 6.60218 5.5 7 5.5C7.55228 5.5 8 5.05228 8 4.5C8 3.94772 7.55228 3.5 7 3.5C6.07174 3.5 5.1815 3.86875 4.52513 4.52513C3.86875 5.1815 3.5 6.07174 3.5 7C3.5 7.55228 3.94772 8 4.5 8C5.05228 8 5.5 7.55228 5.5 7C5.5 6.60218 5.65804 6.22064 5.93934 5.93934Z"
            fill="#FFD600"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div>
        <span
          class="cm-departure-live-departures emotion-4"
        >
          In 
          4, 9, 15
           min
        </span>
         
        <span
          class="cm-departure-live-origin emotion-5"
        >
          from 
          Stratford
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`RouteListItemDepartures live: single departure display correctly 1`] = `
.emotion-0 {
  font-size: 0.8125em;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-top: 14px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: flex-start;
  align-self: flex-start;
  margin-right: 4px;
}

.emotion-2 svg {
  width: 8px;
  height: 8px;
}

.emotion-3 {
  width: 24px;
  height: 24px;
}

.emotion-5 {
  color: #727780;
}

<div>
  <div
    class="cm-route-departures emotion-0"
    data-departure-type="live"
    data-is-live="true"
  >
    <div
      class="cm-departure-live-root emotion-1"
    >
      <div
        class="cm-departure-live-blip emotion-2"
      >
        <svg
          class="emotion-3"
          fill="none"
          viewBox="0 0 8 8"
        >
          <path
            clip-rule="evenodd"
            d="M3.46447 3.46447C4.40215 2.52678 5.67392 2 7 2C7.55228 2 8 1.55228 8 1C8 0.447715 7.55228 0 7 0C5.14348 0 3.36301 0.737498 2.05025 2.05025C0.737498 3.36301 0 5.14348 0 7C0 7.55228 0.447715 8 1 8C1.55228 8 2 7.55228 2 7C2 5.67392 2.52678 4.40215 3.46447 3.46447ZM5.93934 5.93934C6.22064 5.65804 6.60218 5.5 7 5.5C7.55228 5.5 8 5.05228 8 4.5C8 3.94772 7.55228 3.5 7 3.5C6.07174 3.5 5.1815 3.86875 4.52513 4.52513C3.86875 5.1815 3.5 6.07174 3.5 7C3.5 7.55228 3.94772 8 4.5 8C5.05228 8 5.5 7.55228 5.5 7C5.5 6.60218 5.65804 6.22064 5.93934 5.93934Z"
            fill="#FFD600"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div>
        <span
          class="cm-departure-live-departures emotion-4"
        >
          In 
          4
           min
        </span>
         
        <span
          class="cm-departure-live-origin emotion-5"
        >
          from 
          Stratford
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`RouteListItemDepartures scheduled: single departure is displayed correctly 1`] = `
.emotion-0 {
  font-size: 0.8125em;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-top: 14px;
}

.emotion-3.cm-departure-scheduled-statuses--on_time {
  color: #37ab52;
}

.emotion-3.cm-departure-scheduled-statuses--delay {
  color: #d22020;
}

.emotion-3.cm-departure-scheduled-statuses--cancellation {
  color: #d22020;
}

.emotion-4 {
  color: #727780;
}

<div>
  <div
    class="cm-route-departures emotion-0"
    data-departure-type="scheduled"
  >
    <div
      class="cm-departure-scheduled-root emotion-1"
    >
      <span
        class="cm-departure-scheduled-departures emotion-2"
      >
        14:00
         (
        <span
          class="cm-departure-scheduled-statuses cm-departure-scheduled-statuses--on_time emotion-3"
        >
          On time
        </span>
        ) 
        Dartford
      </span>
       
      <span
        class="cm-departure-scheduled-origin emotion-4"
      >
        from 
        Stratford
      </span>
    </div>
  </div>
</div>
`;
